kubernetes-client-api/src/test/java/io/fabric8/kubernetes/client/http/AbstractSimultaneousConnectionsTest.java

```diff
@@ -0,0 +1,214 @@
+  //  by any HttpClient implementation
+  private static final int MAX_HTTP_1_CONNECTIONS = 900; // Should be able to at least make 1024
+
+  private RegisteredServerSocketFactory serverSocketFactory;
+  private MockWebServer mockWebServer;
+
+  private HttpClient.Builder clientBuilder;
+
+  @BeforeEach
+  void prepareServerAndBuilder() {
+    serverSocketFactory = new RegisteredServerSocketFactory();
+    mockWebServer = new MockWebServer();
+    mockWebServer.setServerSocketFactory(serverSocketFactory);
+    clientBuilder = getHttpClientFactory().newBuilder();
+  }
+
+  @AfterEach
+  void stopServer() throws IOException {
+    serverSocketFactory.close();
+    mockWebServer.shutdown();
+  }
+
+  protected abstract HttpClient.Factory getHttpClientFactory();
+
+  // TODO: WIP
+  //  https://github.com/fabric8io/kubernetes-client/issues/4193
+  private void withHttp2() throws IOException {
+    mockWebServer.setProtocols(Collections.singletonList(Protocol.H2_PRIOR_KNOWLEDGE));
+    // TODO: remove! this is just a reminder to FIX JettyHttpClientBuilder constructor where the feature flag
+    //  is enabled (disable http2) by default.
+    ((StandardHttpClientBuilder<?, ?, ?>) clientBuilder).preferHttp11 = false;
+    mockWebServer.start();
+  }
+
+  private void withHttp1() throws IOException {
+    mockWebServer.setProtocols(Collections.singletonList(Protocol.HTTP_1_1));
+    mockWebServer.start();
+  }
+
+  @Test
+  public void http1Connections() throws Exception {
+    final ExecutorService httpExecutor = Executors.newCachedThreadPool();
+    final HttpServer httpServer = HttpServer.create(new InetSocketAddress(0), 0);
+    httpServer.setExecutor(httpExecutor);
+    final CountDownLatch activeConnections = new CountDownLatch(MAX_HTTP_1_CONNECTIONS);
+    httpServer.createContext("/http", exchange -> {
+      activeConnections.countDown();
+      try {
+        activeConnections.await(20, TimeUnit.SECONDS);
+        exchange.sendResponseHeaders(204, 0);
+      } catch (Exception ignore) {
+        // NO OP
+      }
+    });
+    httpServer.start();
+    try (final HttpClient client = clientBuilder.build()) {
+      final List<HttpResponse<AsyncBody>> asyncResponses = new ArrayList<>();
+      final HttpRequest request = client.newHttpRequestBuilder()
+          .uri(String.format("http://localhost:%s/http", httpServer.getAddress().getPort()))
+          .build();
+      for (int it = 0; it < MAX_HTTP_1_CONNECTIONS; it++) {
+        client.consumeBytes(request, (value, asyncBody) -> asyncBody.consume())
+            .whenComplete((r, e) -> asyncResponses.add(r));
+      }
+      assertThat(activeConnections.await(20, TimeUnit.SECONDS)).isTrue();
+      assertThat(asyncResponses).extracting(HttpResponse::code).containsOnly(204);
+    } finally {
+      httpServer.stop(0);
+      httpExecutor.shutdownNow();
+    }
+    assertThat(serverSocketFactory.activeConnections()).isZero();
```

Comment:
Isn't this assertion problematic for JDK?

Reply:
Haven't implemented the JDK tests yet. If it fails, maybe further adjusting the predicate might work. I'll check

Reply:
OK, this one should be removed, thought it was part of the WS test. This is legacy from the previous test implementation.